### PR TITLE
Monitor subprocesses to avoid zombies

### LINKED ISSUE #16410: Process reconciliation as a plugin
### Outline & Motivation

The DDP strategy and its subclasses have a feature called "deadlock detection and process reconciliation". It can ensure that all processes terminate properly when an error occurs on a subset of the ranks. Without this feature, the processes where no errors occur would continue to run and hang/wait at the collectives. 

Pro:
- Can save you costs when running in the cloud.
- No zombie processes you have to manually kill

Con:
- Implementation is hardcoded into the DDPStrategy, does not work well with inheritance
- Makes [trainer exception handling](https://github.com/Lightning-AI/lightning/blob/fc195b95405e9e2629466e5b28c6a9243209d596/src/pytorch_lightning/trainer/call.py#L57-L59) complex 
- Assumes a shared filesystem

### Pitch

1. Remove the feature from PL strategies (#16204)
2. Introduce it as a plugin under fabric
3. Introduce `Strategy.on_exception` that the exception handler can call in a standardized way
4. Re-introduce it in PL strategies once flattening preparations for Fabric integration are done
- [ ] Subtask: Add the plugin to other strategies such as deepspeed:  #16518

A strategy can enable the plugin like so:
```py
if not torch_greater_equal_foo and has_shared_filesystem:
    enable_plugin()
```

and by implementing 
```py
def on_exception(self, exception):
    self.plugin.reconciliate_processes(...)
```

### Additional context

Credit for the ideas @carmocca 

_No response_

cc @justusschock @awaelchli @carmocca

### LINKED ISSUE #16518: DeepSpeed Training Deadlock
### Bug description

I'm using PyTorch Lightning combined with the DeepSpeed strategy (stage 2) to train on 8 V100 GPUs on a single node and I'm running into the following deadlock issue. If I get an OOM on one of the GPUs that does not correspond to the main process that started the job, then that "secondary process" will terminate but all other processes will end up waiting on it (i.e., 100% GPU utilization and nothing happening). This seems like a deadlock issue that lightning should be managing but I'm not sure where that is supposed to be handled and how. Aside from a potential resolution, pointers to how lightning is meant to handle this kind of deadlock and where that happens in the codebase would be super helpful. Thank you!

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component: Trainer
#- PyTorch Lightning Version: 1.9.0
#- PyTorch Version: 1.13.0
#- Python version: 3.7
#- OS: Ubuntu 20
#- CUDA version: 11.6
#- GPU models and configuration: 8 V100s with 32GB each
```

</details>

cc @borda @awaelchli @justusschock

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

This PR adds a process observer thread that monitors the child processes launched with the `SubprocessScriptLauncher`. The observer periodically checks whether the processes have terminated or not. If any processes have non-zero exit code, the thread will forcefully shutdown all other processes. Note that this only quality-of-life improvement applies only to our special process launcher from Lightning for single-node use. For multi-node training, we still rely on the external process launcher (SLURM, torchelastic, etc.) to handle this.

Fixes #16410 
Fixes #16518

Here is a simple example script if you want to test it yourself. It crashes on rank 1 and hangs on rank 0. On this branch, the processes will be cleaned up properly.

```py
import torch
from torch.utils.data import DataLoader, Dataset

from lightning.pytorch import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        if batch_idx > 5 and self.global_rank == 3:
            raise RuntimeError("Rank 2 exit")
        return {"loss": loss}

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
model = BoringModel()
trainer = Trainer(
    max_epochs=100,
    devices=4,
    accelerator="cpu",
    strategy="ddp",
)
trainer.fit(model, train_dataloaders=train_data)
```

cc @borda @carmocca @justusschock @awaelchli

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: src/lightning/fabric/strategies/launchers/subprocess_script.py [hunk 1]
```diff
@@ -159,3 +170,53 @@ def _hydra_subprocess_cmd(local_rank: int) -> Tuple[Sequence[str], str]:
...
+        if all(return_code == 0 for return_code in return_codes):
+            return True
+
+        for proc in self._child_processes:
+            if proc.returncode:
+                _logger.info(
+                    f"Child process with PID {proc.pid} terminated with code {proc.returncode}."
+                    f" Forcefully terminating all other processes to avoid zombies 🧟"
+                )
+                self._terminate_all()
```
Comment (carmocca) [inline src/lightning/fabric/strategies/launchers/subprocess_script.py:213]: Doesn't this partially repeat what's in https://github.com/Lightning-AI/lightning/blob/master/src/lightning/pytorch/trainer/connectors/signal_connector.py#L107-L115 already?

Comment (awaelchli) [reply src/lightning/fabric/strategies/launchers/subprocess_script.py:213]: It is conceptually different. This logic in the signal connector ensures that all processes are terminated when a SIGTERM arrives. This means the process is being terminated in a regular fashion by an external signal, and we are just ensuring all (child) processes get terminated completely.

Here however, we are monitoring on the side if any of the processes exit unexpectedly while others are still running. This prevents a stall of the program. 

Adding it to the multiprocessing launcher is also possible, the logic will be slightly different.

Comment (carmocca) [reply src/lightning/fabric/strategies/launchers/subprocess_script.py:213]: Couldn't this also include the logic to terminate subprocesses if it's the main process that gets killed (normally by a signal)? Consolidating this logic together would be useful

Comment (awaelchli) [reply src/lightning/fabric/strategies/launchers/subprocess_script.py:213]: That doesn't fit well into a thread. 1) there is no need to have a thread if the main process can kill the processes that it have launched. 2) If the main process gets terminated, the thread would also be terminated and so it wouldn't be able to run its logic. I don't see a need to consolidate or abstract this.

**Pull request merged**